design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,493 @@
+	ReasonRejectLUNVolumes               = "LUN disks aren't supported to be migrated yet"
+	ReasonRejectedMultipleVMIs           = "Volumes need to belong to the same VMI"
+	ReasonRejectedPending                = "Volumes need to belong to a running VMI"
+	ReasonRejectedMultipleVMIsAndPending = "Volumes need to belong to the same running VMI"
+)
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhasePending    VolumeMigrationPhase = "Pending"
+	VolumeMigrationPhaseScheduling VolumeMigrationPhase = "Scheduling"
+	VolumeMigrationPhaseRunning    VolumeMigrationPhase = "Running"
+	VolumeMigrationPhaseSucceeded  VolumeMigrationPhase = "Succeeded"
+	VolumeMigrationPhaseFailed     VolumeMigrationPhase = "Failed"
+	VolumeMigrationPhaseUnknown    VolumeMigrationPhase = "Unknown"
+)
+
+type MigratedVolumeValidation string
+
+const (
+	MigratedVolumeValidationValid    MigratedVolumeValidation = "Valid"
+	MigratedVolumeValidationPending  MigratedVolumeValidation = "Pending"
+	MigratedVolumeValidationRejected MigratedVolumeValidation = "Rejected"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type VolumeMigrationState struct {
+	MigratedVolume `json:",inline"`
+	Validation     MigratedVolumeValidation `json:"validation,omitempty"`
+	Reason         *string                  `json:"reason,omitempty"`
+}
+
+type VolumeMigrationStatus struct {
+	VolumeMigrationStates       []VolumeMigrationState                    `json:"volumeMigrationStates,omitempty"`
+	VirtualMachineInstanceName  *string                                   `json:"virtualMachineInstanceName,omitempty"`
+	VirtualMachineMigrationName *string                                   `json:"virtualMachineMigrationName,omitempty"`
+	StartTimestamp              *metav1.Time                              `json:"startTimestamp,omitempty"`
+	EndTimestamp                *metav1.Time                              `json:"endTimestamp,omitempty"`
+	Phase                       VolumeMigrationPhase                      `json:"phase,omitempty"`
+	PhaseTransitionTimestamps   []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
+}
+```
+
+Users may define more than one migrated volume, and those volumes may be associated with the same VMI, a different VMI, or none at all. The volumes defined on the same VMI will be categorized and grouped by the volume migration controller.
+
+At the moment, PVCs can only be migrated using this API if they belong to a single running virtual machine. Consequently, all volume migration objects that have migrated volumes that are associated with different VMIs will be rejected by the controller. The object is in the `Pending` state if none of the migrated volumes have been assigned to a VMI, since the VMI might be created at a later time.
```

Comment:
Do you intend to support volumes owned by different VMIs and any PVC in general in the future?

Reply:
Maybe, my first version of this CRD was supporting multiple VMIs and PVCs.  However, after discussing this with @awels and @vasiliy-ul we decided to keep it simple for the first version and to support a single VM per volume migration object.
Additionally, @awels is working on an overlaying tool that should create multiple volume migrations object based on VMs and that would take care to create the destination volumes.